src/providers/base_javascript.js

```diff
@@ -0,0 +1,211 @@
+		let rootName = depsObject["name"]
+		let rootVersion = depsObject["version"]
+		if (!rootVersion) {
+			rootVersion = defaultVersion
+		}
+		let mainComponent = this.#toPurl(rootName, rootVersion);
+
+		let sbom = new Sbom();
+		sbom.addRoot(mainComponent)
+
+		let dependencies = depsObject["dependencies"] || {};
+		this.#addAllDependencies(sbom, sbom.getRoot(), dependencies)
+		let packageJson = fs.readFileSync(manifest).toString()
+		let packageJsonObject = JSON.parse(packageJson);
+		if (packageJsonObject.exhortignore !== undefined) {
+			let ignoredDeps = Array.from(packageJsonObject.exhortignore);
+			sbom.filterIgnoredDeps(ignoredDeps)
+		}
+		return sbom.getAsJsonString(opts)
+	}
+
+	/**
+   * This function recursively build the Sbom from the JSON that npm listing returns
+   * @param sbom this is the sbom object
+   * @param from this is the current component in bom (Should start with root/main component of SBOM) for which we want to add all its dependencies.
+   * @param dependencies the current dependency list (initially it's the list of the root component)
+   * @private
+   */
+	#addAllDependencies(sbom, from, dependencies) {
+		Object.entries(dependencies)
+			.filter(entry => entry[1].version !== undefined)
+			.forEach(entry => {
+				let [name, artifact] = entry;
+				let purl = this.#toPurl(name, artifact.version);
+				sbom.addDependency(from, purl)
+				let transitiveDeps = artifact.dependencies
+				if (transitiveDeps !== undefined) {
+					this.#addAllDependencies(sbom, sbom.purlToComponent(purl), transitiveDeps)
+				}
+			});
+	}
+
+	#executeListCmd(includeTransitive, manifestDir) {
+		const listArgs = this._listCmdArgs(includeTransitive, manifestDir);
+		try {
+			return invokeCommand(this._cmdName(), listArgs)
+		} catch (error) {
+			throw new Error(`failed to list dependencies via "${this._cmdName()} ${listArgs.join(' ')}" - Error: ${error}`, {cause: error});
+		}
+	}
+
+	#version() {
+		try {
+			invokeCommand(this._cmdName(), ['--version'], {stdio: 'ignore'});
+		} catch (error) {
+			if (error.code === 'ENOENT') {
+				throw new Error(`${this._cmdName()} is not accessible`);
+			}
+			throw new Error(`failed to check for package manager binary at ${this._cmdName()}`, {cause: error})
+		}
+	}
+
+	#createLockFile(manifestDir) {
+		// in windows os, --prefix flag doesn't work, it behaves really weird , instead of installing the package.json fromm the prefix folder,
+		// it's installing package.json (placed in current working directory of process) into prefix directory, so
+		let originalDir = process.cwd()
+		if (os.platform() === 'win32') {
+			process.chdir(manifestDir)
+		}
+		const args = this._updateLockFileCmdArgs(manifestDir);
+		try {
+			invokeCommand(this._cmdName(), args)
+		} catch (error) {
+			throw new Error(`failed to create lockfile "${args}" - Error: ${error}`, {cause: error});
+		} finally {
+			if (os.platform() === 'win32') {
+				process.chdir(originalDir)
+			}
+		}
```

Comment:
I wonder does setting `{cwd: manifestDir}` for child_process work here, or do we really have to `process.chdir`?

Reply:
I didn't dare to change it in this PR to avoid re-testing on different platforms.